libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/memory_pools.hpp

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <eosio/chain/webassembly/eos-vm-oc/memory.hpp>
+#include <eosio/chain/sync_call_resource_pool.hpp>
+
+namespace eosio::chain::eosvmoc {
+
+/*
+ * Manages a set of memory pools for sync calls.
+ * Shallow call depths use dedicated memory pools with decreasing number
+ * of pre-allocated slices. Deep call depths share one pool with one
+ * pre-allocated * memory slice.
+ *
+ * Depth 1 is allocated highest slices, then each depth decrements by one,
+ * until reaching one slice for all remaining depths.
+ *
+ * For example, if max call depth is 10 and we allocate 5 slices for the first
+ * depth, we will have 5 memory pools:
+ *
+ * pool 1: 5 slices, for depth 1
+ * pool 2: 4 slices, for depth 2
+ * pool 3: 3 slices, for depth 3
+ * pool 4: 2 slices, for depth 4
+ * pool 5: 1 slice,  for depths 5 to 10
+ *
+ * */
+class memory_pools {
+   public:
+      memory_pools();
+
+      // Acquire memory from the pool of `call_depth`
+      eosvmoc::memory* acquire_mem(uint32_t call_depth);
+
+      // Release memory `m` back to the pool of `call_depth`
+      void release_mem(uint32_t call_depth, eosvmoc::memory* m);
```

Comment:
I wonder if it makes any sense for `memory` to be queryable for its slices so you don't need to parrot back the call_depth -- could just query `memory` for its slice count and know what pool to place it back it. Not really a big deal to parrot back call_depth though

Reply:
Do you want me to put a `sliced_pages` member to `memory` and query it in `release_mem` so we don't need `call_depth` parameter? I think it is cleaner and safer (not put back a memory into a wrong pool by accident)

Reply:
I don't think it's necessary since we declare the scoped_exit so closely nearby to the aquire_mem, so it's up to you

Reply:
OK. I will keep it as is.